### ISSUE TITLE
Proguard 7.0.1 -keepkotlinmetadata prints Warning: Kotlin metadata errors encountered XXX  Type has no reference for its type alias

### ISSUE THREAD

**Issue by lsuski** [Guardsquare/proguard#94]
Hi, 
I've tested fix for #84 with proguard 7.0.1 and It does not work for me. I have lots of warnings for in my opinion simple cases like
```
Warning: Kotlin metadata errors encountered in xom/example/SomeException. Not processing the metadata for this class.
  Type has no reference for its type alias "kotlin/Exception".
```
or for variables like ` var someVar by someDelegate()`
```
Warning: Kotlin metadata errors encountered in some/Class. Not processing the metadata for this class.
  Property someVar has no backingFieldSignature, getterSignature or setterSignature.
```

Although obfuscation does not fail kotlin metadata are not kept for those classes

**Comment by mrjameshamilton** [comment]
Hi @lsuski could you provide a sample some samples for these cases? Then we can look into them.

We do notice sometimes there are Kotlin metadata problems or something is missing, even in the Kotlin libraries themselves. If there is a problem with the Kotlin metadata we throw the metadata away rather than write out broken metadata.

For keeping Kotlin metadata we have taken the approach of only keeping the metadata attached to a class if

* `-keepkotlinmetadata`
* the class or it's members are explicitly kept

The reasoning behind this is that in many cases the Kotlin metadata is not actually required and it just increases the size of the app. The two important cases where Kotlin metadata are required are in libraries and reflection.

Regarding delegate properties - you're right, this is not yet supported.

**Comment by lsuski** [comment]
Ok, so in my case there is library. I'll try to prepare simple sample because I can't publish my own project

**Comment by lsuski** [comment]
[ProguardKotlinMetadataSample.zip](https://github.com/Guardsquare/proguard/files/5493664/ProguardKotlinMetadataSample.zip)

This is sample project. I think tha problem is with kotlin typealiases. If I have file with reference to `kotlin.Exception` than there is error in processing metadata

**Comment by mrjameshamilton** [comment]
I've investigated your sample and I can confirm there is a problem when processing libraries with Kotlin metadata. I have managed to come up with some workarounds though. I've added the Kotlin classes as `-injars` and added a `-keep` rule for them, and filter them out during the output writing. This seems to work. Can you confirm?

[WorkaroundProguardKotlinMetadataSample.zip](https://github.com/Guardsquare/proguard/files/5497875/WorkaroundProguardKotlinMetadataSample.zip)

**Comment by lsuski** [comment]
I've made little error in sample: `libraryjars(fileTree("libs_not_obfuscated/source_repository"))` should be `libraryjars(fileTree("libs_not_obfuscated"))`.

Your fix seems to work, I've tested it in my project (much more complicated with lots of dependencies), I've needed to add
```
-dontwarn kotlin.**
-dontwarn kotlinx.**
```
bacause I have libraryjars written in kotlin

**Comment by lsuski** [comment]
I stiil have some errors like ` Default implementations class name does not end with $DefaultImpls`

**Comment by mrjameshamilton** [comment]
Yes, you're right. Normally, the kotlin library jars should but `-libraryjars` inputs but it seems metadata library classes is not yet supported (since the alias is in the library, ProGuard needs to load the metadata to know this). So I changed them to program `-injars`: `injars configurations.kotlinLibs, filter: '**.class'`

You should be able to remove the library jar version, as this will give you duplicate class errors. I'll create an issue to fix the problem with the Kotlin library metadata.

Can you open a new issue with a sample for the `$DefaultImpls` problem?

**Comment by lsuski** [comment]
Ok, I'll try to reproduce `DefaultImpls` warning in sample
